ns1/datadog_checks/ns1/check.py

```diff
@@ -0,0 +1,448 @@
+import errno
+import json
+import logging
+import os
+
+from requests.exceptions import ConnectionError, HTTPError, InvalidURL, Timeout
+
+from datadog_checks.base import AgentCheck, ConfigurationError
+
+from .ns1_url_utils import Ns1Url
+
+
+class Ns1Check(AgentCheck):
+    NS1_SERVICE_CHECK = "ns1.can_connect"
+
+    def __init__(self, name, init_config, instances):
+        super(Ns1Check, self).__init__(name, init_config, instances)
+        self.api_endpoint = self.instance.get("api_endpoint")
+        if not self.api_endpoint:
+            raise ConfigurationError('NS1 API endpoint must be specified in configuration')
+
+        self.api_key = self.instance.get("api_key")
+        if not self.api_key:
+            raise ConfigurationError('NS1 API key must be specified in configuration')
+
+        self.headers = {"X-NSONE-Key": self.api_key}
+
+        self.metrics = self.instance.get("metrics")
+        if not self.metrics or len(self.metrics) == 0:
+            raise ConfigurationError('Invalid metrics config!')
+
+        self.query_params = self.instance.get("query_params")
+        self.usage_count_path = "/opt/datadog-agent/log"
```

Comment:
Writing to a file between check runs is very anti-pattern. It's not going to work on other OS, and it leads to concurrency issues when you have two instances of the ns1 check running on the same agent.

It looks like you only use that to keep track of the previous metric values so that you compute a diff. You can use `self.monotonic_count` instead of `self.count` and the core agent will handle the computation for you.
In case you need to store some state between check runs, you can use an instance variable in this class.

Reply:
Thank you for review.
I completely agree with your comment, however this was done out of necessity - NS1 API returns count of queries, decisions etc. as array of tuples with (<timestamp>,<count>). Each count is separated in 'buckets' for each period, and at the start of new time period goes back to 0 and increase from there for subsequent calls as count increases.
So I keep track of both time period and count for each metric in file to preserve it between agent restarts.
I considered monotonic_count, but as value goes down to 0 when new 'bucket' is formed, it won't work properly (or at least that's my understanding how monotonic_count works)

This workaround does works and it's tested on linux/windows, wasn't tested on mac.
Regarding multiple instance of the check running on the same agent - we will document that if that is needed for whatever reason, instances should not be checking same metrics as it will lead to duplicated counts for the same metric.

If this is a hard no-no, then we will have to either drop all count metrics which will render integration mostly useless as most of the metrics are count, or to tell users that after agent restart all counters will experience huge spikes in value, which ultimately leads to questionable value of the integration.

Is there any way to submit count value to datadog with both timestamp and value? That could potentially be a solution.

Reply:
If you need to use a file, rather than harcoding the path the suggestion is to use the persistent cache that is provided for checks https://github.com/DataDog/integrations-extras/pull/883.

Reply:
I removed the data being stored in the file to the cache that is already implemented